### PR TITLE
Ensure native histograms counter reset hints are corrected when merging results from different sources

### PR DESCRIPTION
**Pull request by fionaliao**
#### What this PR does
When merging samples from different iterators, histogram counter reset hints need to be recalculated. The `mergeIterator` currently does not do so. This PR fixes that by remembering the iterator for a sample in a batch, so when the next sample comes along, the iterators for the two samples can be compared and an unknown counter reset hint is set for the next sample if the iterators don't match. See https://github.com/prometheus/prometheus/issues/15346 for related discussion. 

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/querier/batch/stream.go [hunk 1]
```diff
@@ -142,28 +163,53 @@ func (bs *batchStream) merge(batch *chunk.Batch, size int) {
...
 			bs.next()
 		} else if t1 > t2 {
-			populate(batch, rt)
+			populate(batch, rt, fromIncomingBatch)
 			batch.Next()
 		} else {
 			if (rt == chunkenc.ValHistogram || rt == chunkenc.ValFloatHistogram) && lt == chunkenc.ValFloat {
 				// Prefer histograms than floats. Take left side if both have histograms.
-				populate(batch, rt)
+				populate(batch, rt, fromBatchStream)
```
Comment (krajorama) [inline pkg/querier/batch/stream.go:210]:
```suggestion
				populate(batch, rt, fromIncomingBatch)
```

File: pkg/querier/batch/stream.go [hunk 2]
```diff
@@ -142,28 +163,53 @@ func (bs *batchStream) merge(batch *chunk.Batch, size int) {
...
 			} else {
-				populate(bs.curr(), lt)
+				populate(bs.curr(), lt, fromIncomingBatch)
```
Comment (krajorama) [inline pkg/querier/batch/stream.go:212]:
```suggestion
				populate(bs.curr(), lt, fromBatchStream)
```

**Comment by fionaliao** [comment]
Implementation and testing is done (pending CI), but I want to extract the mimir-prometheus update into a separate PR before removing the "WIP"

**Comment by fionaliao** [comment]
mimir-prometheus update was done as part of https://github.com/grafana/mimir/pull/10033

**Review by krajorama — COMMENTED** [review]
looking good a few nits added first round

File: pkg/storage/chunk/prometheus_chunk.go [hunk 3]
```diff
@@ -119,6 +119,8 @@ func (p *prometheusHistogramChunk) AddFloatHistogram(_ int64, _ *histogram.Float
 // AddHistogram adds another histogram to the chunk. While AddHistogram works, it is only implemented to make tests
 // work, and should not be used in production. In particular, it appends all histograms to single chunk, and uses new
 // Appender for each invocation.
+// AddHistogram adds another histogram to the chunk. While AddHistogram works, it is only implemented to make tests
```
[the remark this answers is not in the document]
Comment (fionaliao) [reply pkg/storage/chunk/prometheus_chunk.go]: Removed old comment in https://github.com/grafana/mimir/pull/9909/commits/183f987d2e7f0131691cfef293214a1e985c5437

**Review by krajorama — APPROVED** [review]

**Pull request merged**